### DISCUSSION TITLE
OAuthError: invalid_client :  The client MUST NOT use more than one authentication method in each request

### DISCUSSION THREAD

**Discussion by mandeeplohan** [apache/superset#17993] [Q&A / Help]
We configured the Superset_config.py as suggested on the documentation page and a custom sso as suggested in documentation . It is getting authenticated using OIDC and returning with a valid code. However it is not proceding after that. Error is 

authlib.integrations.base_client.errors.OAuthError: invalid_client: The client MUST NOT use more than one authentication method in each request.
superset_app            | 2021-12-23 14:19:50,024:ERROR:superset.views.base:invalid_client: The client MUST NOT use more than one authentication method in each request.
superset_app            | Traceback (most recent call last):

Changes in Superset_cofig.py file*****

from custom_sso_security_manager import CustomSsoSecurityManager
from flask_appbuilder.security.manager import AUTH_OAUTH,AUTH_OID

CUSTOM_SECURITY_MANAGER = CustomSsoSecurityManager
AUTH_TYPE = AUTH_OAUTH

OAUTH_PROVIDERS = [
{ 'name':'PING',
'token_key':'access_token', # Name of the token in the response of access_token_url
'icon':'fa-address-card', # Icon for the provider
'remote_app': {
'client_id':'client_id', # Client Id (Identify Superset application)
'client_secret':'client_secret', # Secret for this Client Id (Identify Superset application)
'client_kwargs':{
'scope': 'openid profile address email phone', # Scope for the Authorization
},
'access_token_method':'POST', # HTTP Method to call access_token_url
'access_token_params':{ # Additional parameters for calls to access_token_url
'client_id':'client_id',
'grant_type' : 'authorization_code',
'client_secret' :'client_secret',
'redirect_uri' : 'http://url-blaa-bla:8088/'
},
'access_token_headers':{ # Additional headers for calls to access_token_url
'Authorization': 'Basic Base64EncodedClientIdAndSecret',
'Content-Type': 'application/x-www-form-urlencoded'
},
'api_base_url':'api_base_url',
'access_token_url':'access_token_url',
'authorize_url':'authorize_url'
}
}
]

AUTH_USER_REGISTRATION = True
AUTH_USER_REGISTRATION_ROLE = "Public"

#***********

Actual results
Authentication happening with Get call and returing with a valid code and state on the browser.

what actually happens.

http://url-bla-bla:8088/

Eneter MS ID and password

Auth success

in browser it comes with a valid code:
https://url-bla-bla:8088/oauth-authorized/PING?code=aaaaaaaaaaaaaaaaa&state=bbbbbbbbbbb.cccccccccc.ddddddd-ffffff-gggggggg-Y

Error on browser is "invalid redirect uri

LOGS:
superset_app | 2021-09-20 14:45:19,097:DEBUG:authlib.integrations.base_client.base_app:Saving authorize data: {'redirect_uri': 'http://localhost:8088/oauth-authorized/PING', 'nonce': 'S6JvBApadi4z3wOIyMWE', 'url': 'https://url/as/authorization.oauth2?response_type=code&client_id=client_id&redirect_uri=http%3A%2F%2Flocalhost%3A8088%2Foauth-authorized%2FPING&scope=openid+profile+address+email+phone&state=aaaaaaa.bbbbbbb.cccccc-ddddd-eeeee-Y&nonce=S6JvBApadi4z3wOIyKVU', 'state': 'eyL0eEAiOiMNS1QiLCJhbGciOiJIUzI1NiJ9.eeeeeee.bbbbbb-vvvvv-vvvvvv-Y'}
superset_app | 172.18.0.1 - - [20/Sep/2021:14:45:19 +0000] "GET /login/PING?next= HTTP/1.1" 302 951 "http://localhost:8088/login/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/93.0.4577.82 Safari/537.36"
superset_app | 127.0.0.1 - - [20/Sep/2021:14:45:22 +0000] "GET /health HTTP/1.1" 200 2 "-" "curl/7.64.0"

How to reproduce the bug
Git clone: https://github.com/apache/superset.git
configure superset_config.py with Oauth changes
add a custum sso file as defined in documention
build docker image : docker build -t superset-dev:latest .
docker-compose -f docker-compose-non-dev.yml up
http://localhost:8088/
enter MS id and password
Auth success
in browser it comes with a valid code:
https://localhost:8088/oauth-authorized/PING?code=aaaaaaaaaaaaaaaaa&state=bbbbbbbbbbb.cccccccccc.ddddddd-ffffff-gggggggg-Y
Error on browser is "This site can’t be reached"
Environment
Local : https://localhost:8088/

(please complete the following information):

File "/usr/local/lib/python3.8/site-packages/authlib/integrations/requests_client/oauth2_session.py", line 117, in handle_error
superset_app            |     raise OAuthError(error_type, error_description)
superset_app            | authlib.integrations.base_client.errors.OAuthError: invalid_client: The client MUST NOT use more than one authentication method in each request.
superset_app            | 2021-12-23 14:19:50,024:ERROR:superset.views.base:invalid_client: The client MUST NOT use more than one authentication method in each request.
superset_app            | Traceback (most recent call last):
superset_app            |   File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
superset_app            |     rv = self.dispatch_request()
superset_app            |   File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
superset_app            |     return self.view_functions[rule.endpoint](**req.view_args)
superset_app            |   File "/usr/local/lib/python3.8/site-packages/flask_appbuilder/security/views.py", line 659, in oauth_authorized
superset_app            |     resp = self.appbuilder.sm.oauth_remotes[provider].authorize_access_token()
superset_app            |   File "/usr/local/lib/python3.8/site-packages/authlib/integrations/flask_client/remote_app.py", line 76, in authorize_access_token
superset_app            |     token = self.fetch_access_token(**params)
superset_app            |   File "/usr/local/lib/python3.8/site-packages/authlib/integrations/base_client/remote_app.py", line 112, in fetch_access_token
superset_app            |     token = client.fetch_token(token_endpoint, **kwargs)
superset_app            |   File "/usr/local/lib/python3.8/site-packages/authlib/oauth2/client.py", line 203, in fetch_token
superset_app            |     return self._fetch_token(
superset_app            |   File "/usr/local/lib/python3.8/site-packages/authlib/oauth2/client.py", line 226, in _fetch_token
superset_app            |     return self.parse_response_token(resp.json())
superset_app            |   File "/usr/local/lib/python3.8/site-packages/authlib/oauth2/client.py", line 380, in parse_response_token
superset_app            |     self.handle_error(error, description)
superset_app            |   File "/usr/local/lib/python3.8/site-packages/authlib/integrations/requests_client/oauth2_session.py", line 117, in handle_error
superset_app            |     raise OAuthError(error_type, error_description)
superset_app            | authlib.integrations.base_client.errors.OAuthError: invalid_client: The client MUST NOT use more than one authentication method in each request.
superset_app            | 10.175.238.7 - - [23/Dec/2021:14:19:50 +0000] "GET /oauth-authorized/PING?code=GYAb-14DB_65NwWzDSQiw4AUzeGpvc04Oasasasasas&state=eyJ0eXAiOiJKV1QiLCJhbGciqwqwwweeUzI1NiJ9.eyJuZXh0IjpbIiJdfQ.fdfto7RZFzoca-sdsdsdsd-jsGjhSPMH_ltWoj85-Y HTTP/1.1" 500 0 "https://hats-superset.mandeep.com/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36"

**Comment by mandeeplohan** [comment]
@eschutho @dpgaspar @junlincc  Any help here please? I am trying since 2 months on Oauth and stuck. I am not getting any help anywhere and even from superset team. I know its opensource but some help would really do it. If this is not resolved management has asked to sidelined the superset explore something else :(  because i have already wasted 2 months on just SSO and posting questions here since 2 months. :( :(

**Comment by mandeeplohan** [comment]
@amitmiran137 @dpgaspar @eschutho Can i expect any help here?

**Comment by amitmiran137** [comment]
I'm sorry I'm not familiar with that side.
I believe this is mostly relies on FAB realm

**Comment by eschutho** [comment]
Hi @mandeeplohan I'm not familiar with this part of the app either. Most of this functionality lies outside of Superset, like Amit mentioned, but I'd suggest taking a look at the source of the error code from the oauth documentation: https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-31#section-2.3

My first guess would be that you have the default `AUTH_USER_REGISTRATION_ROLE` set to "Public" when that role shouldn't require authentication. Have you tried setting that config to a different role like Admin?

**Comment by mandeeplohan** [comment]
@amitmiran137 @eschutho  Thanks for your time and response. This would help me for sure.  I have one question based on your response.
1. How this functionality lies outside of Superset?  Superset is providing the way to configure Oauth in the Superset_config.py file.  [https://superset.apache.org/docs/installation/configuring-superset](url) 

I think somewhere Superset is using some library which user dont have access to modify code.

I am good with superset and many charts has been created easily. However, i am stuck on last part I.e.  to enable OAuth or LDAP.  Management does not want separate  creds which we can create easily in superset. 

I dont know what should i do.  I was following this link : https://superset.apache.org/docs/installation/configuring-superset

**Comment by Chinyeins** [comment]
> Error on browser is "invalid redirect uri

Hey, idk if this helps you out, but you said the error code is redirect uri invalid. This means that when the auth and token requests are being processed, the incoming parameters of eacht request are being checked for the required params needed in each rq. Since it says invalid redirect uri, there might be a mismatching redirect uri problem.

Authorized Endpoint of superset: (https://superset.apache.org/docs/installation/configuring-superset/)

`https://<superset-webserver>/oauth-authorized/PING`


this redirect uri has to match your client configuration.

**Comment by assofour** [comment]
I have very similar issue as the log 

2023-02-28 03:38:11,822:DEBUG:authlib.integrations.base_client.base_app:Saving authorize data: {'redirect_uri': 'http://xxxxxxxxx', 'url': 'https://xxxxxxxxx"

However, the redirect_uri has been set to https explicitly and 

ENABLE_PROXY_FIX = True
PROXY_FIX_CONFIG = {"x_for": 1, "x_proto": 1, "x_host": 1, "x_port": 0, "x_prefix": 1} 

in Superset_config.py, but still remain the http in real request.
